core/modules/replica/Configuration.cc

```diff
@@ -23,818 +23,1732 @@
+        throw logic_error(_context(__func__) + " worker: '" + name +"' is already disabled.");
+    }
+    workerJson["is_enabled"] = 0;
+    return _updateWorker(lock, WorkerInfo(workerJson));
 }
 
 
-WorkerInfo Configuration::setWorkerDbUser(string const& name,
-                                          string const& user,
-                                          bool updatePersistentState) {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->setWorkerDbUser(name, user, updatePersistentState);
+WorkerInfo Configuration::updateWorker(WorkerInfo const& info) {
+    util::Lock const lock(_mtx, _context(__func__));
+    _assertWorkerExists(lock, info.name, true);
+    return _updateWorker(lock, info);
 }
 
 
-WorkerInfo Configuration::setWorkerLoaderHost(string const& name,
-                                              string const& host,
-                                              bool updatePersistentState) {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->setWorkerLoaderHost(name, host, updatePersistentState);
+json Configuration::toJson(bool showPassword) const {
+    util::Lock const lock(_mtx, _context(__func__));
+    return _toJson(lock, showPassword);
 }
 
 
-WorkerInfo Configuration::setWorkerLoaderPort(string const& name,
-                                              uint16_t port,
-                                              bool updatePersistentState) {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->setWorkerLoaderPort(name, port, updatePersistentState);
+json Configuration::_toJson(util::Lock const& lock, bool showPassword) const {
+    return _data;
 }
 
-
-WorkerInfo Configuration::setWorkerLoaderTmpDir(string const& name,
-                                                string const& tmpDir,
-                                                bool updatePersistentState) {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->setWorkerLoaderTmpDir(name, tmpDir, updatePersistentState);
+json const& Configuration::_get(
+        util::Lock const& lock, string const& category, string const& param) const
+{
+    json::json_pointer const pointer("/" + category + "/" + param);
+    if (!_data.contains(pointer)) {
+        throw invalid_argument(
+                _context(__func__) + " no such parameter for category: '" + category
+                + "', param: '" + param + "'");
+    }
+    return _data.at(pointer);
 }
 
 
-WorkerInfo Configuration::setWorkerExporterHost(string const& name,
-                                                string const& host,
-                                                bool updatePersistentState) {
-    util::Lock lock(_mtx, context(__func__));
-    return _impl->setWorkerExporterHost(name, host, updatePersistentState);
+json& Configuration::_get(
+        util::Lock const& lock, string const& category, string const& param)
+{
+    return _data[json::json_pointer("/" + category + "/" + param)];
```

Comment:
Is this the same as `_data[category][param]`? Which one is faster?

Reply:
Though I myself didn't have a chance to test and compare the performance of both methods, there is one important advantage of the pointer vs the dictionary key-based access. If either key (or both) isn't present in `_data` the square bracket operator will automatically register the key and make an empty (of type `json::empty`)  `json` object created for the key. This is NOT the desired behavior for the non-const method  `json& Configuration::_get` because the API should not allow registering arbitrary parameters if a client code passes some random strings. In contrast with that, the `json::json_pointer(`-based accessor will throw the exception `json::out_of_range` in case of non-existing keys. Unfortunately, this behavior is not explained in the documentation for the JSON library. I "discovered" it when I was experimenting with the library, and I found a direct confirmation for that later one of the unit tests: https://github.com/nlohmann/json/blob/6f551930e5c7ef397056de121c0da82f77573cca/test/src/unit-json_pointer.cpp   (look for "unresolved access").